README
-- SDL-Ball --
 - A DX-Ball/MegaBall/Breakout clone for linux+friends (and windows)
   with pretty graphics.

-- Index --
1.0 SDL-Ball Info and License
1.1 Objective
1.2 Controls and Keys
1.2.1 Keyboard Controls
1.2.2 Mouse Controls
1.2.3 WiiMote Controls
1.2.4 Joystick Controls
1.3 Saving and Loading
1.4 Configuration
1.4.1 Mouse
1.4.2 Keyboard
1.4.3 Digital Joystick
1.4.4 Analog Joystick
1.4.5 WiiMote
1.4.6 Config options explained

2.0 Compiling
2.1  Without sound
2.2  With WiiMote support
2.3  For system-wide installation

3.0 Themes
3.1 How to use a diffrent theme
3.2 How to make your own theme

4.0 Contributors and Credits

5.0 Frequently Asked Questions (FAQ)

------------------------------------

1.0 SDL-Ball Info
Author: Jimmy Christensen
Project page: https://github.com/DusteDdk/SDL-Ball
Web site: http://sdl-ball.sourceforge.net/

License:
Code, Graphics, Levels, Sound are GPLv3
See themes/default/font/COPYING for font licenses.

2020:
SDL-Ball was a hobby project of mine, it was my first computer game,
and I did learn a lot from it. Lots of things are done wrong here, because,
at the time, I didn't know any better. So don't use this as an example
of how to do a game. That being said, it WORKS, and I'm not going to
rewrite it, or majorly refactor it to make it more "right".
I'm also not going to hide it out of shame, I'm actually proud that I managed
to make it back when I did.. As far as I recall, I wrote most of it in Kate.
So, I'm keeping it up, it's actually a nice game and people would miss it if
I take it down. I'll also try and maintain it a bit better than I've done the
past.. Many years. Other than that, I've nothing to add :)

200?
SDL-Ball is my hobby project, I'm using it to learn C++ and
how not to design software, I learn by trial and error, thats
what makes it fun for me, so please, spare me the lessons.
If you want to clean it up, I'll gladly accept your patches.

------------------------------------

1.1 Objective
The objective of the game is to control the paddle in order to keep
the ball from falling down, by precise controlling you are able to
reflect the ball into bricks. Hitting bricks gives points and powerups.
Collecting powerups make the game easier or harder depending on the type.
Powerups gives you "coins" (you never now exactly how many coins you have),
and you can spend them on... Powerups :D
When all breakable bricks are destroyed, the level ends, and next level is on.
When all levels are completed, game start from level1 again, but with twice the score.

------------------------------------

1.2 Controls and keys
There are several ways to control the paddle, the preferred is using the mouse
but you can also use your keyboard, joystick or wiimote (wiimote, see 1.4.5 2.0 and 2.2)

1.2.1 Keyboard controls
w   - Bring up wiimote connection dialog (only if compiled with wiiuse)
s   - Take Screenshot (saves in ~/.config/sdl-ball/screenshots/)
u   - Turn off screen clearing (gives some really cool/weird effects if background is turned off)
p   - Pause
Pause -Pause
ESC - Pause and Show Menu
F1  - Show titlescreen
F11 - Toggle Fullscreen
c   - Show/Hide clock (Shows what time it is)

v   - Select Powerup (Next)
b   - Buy selected Powerup
n   - Select Powerup (Previous)

Arrow Up - Shoot/Launch ball
Arrow Left and Right - Move paddle left or right (See 1.4.2)


1.2.2 Mouse controls
Button1 - Shoot/Launch ball
Button2(Right Click) - Buy Powerup.
Mouse Down/Up - Select Powerup to buy.
X axis - Move Paddle

1.2.3 WiiMote controls
Button 1 - Buy Item.
Button 2 - Shoot/Launch ball
+ left and right - Select powerup to buy.
Tilt left or right to move paddle.

1.2.4 Joystick controls
Button1 - Shoot/Launch ball
Button2 - Buy powerup.
X axis - Move Paddle
Y axis - Selevt powerup to buy.

------------------------------------

1.3 Saving and Loading
To save your game, press ESC, and click the save-button in the menu
choose a slot and enter a name for the savegame,
then press enter to save.
Saving only saves the game as it is in the START of the level.
You can't save in level 1... Obviously, since you made no progress :)

To load a game, press ESC and click the load-button in the menu
then choose a game.

------------------------------------

1.4 Configuration
~/ means your home-directory.
The game saves its configuration in ~/.config/sdl-ball/settings.cfg
If you choose an unsupported resolution you can correct it in that file, see 1.4.6.
If you want to clear the highscores list, delete ~/.config/sdl-ball/highscores.txt
If you want to clear savegames, delete ~/.config/sdl-ball/savegame.sav
Edit this file with any text-editor to change options that are not reached from the menu.

1.4.1 Mouse
The mouse have no configuration options, you might want to adjust
your mousespeed from your window-manager.

1.4.2 Keyboard
Use the arrow keys to control the paddle, up to shoot and launch the ball.
You can change theese in the config file with the options:
leftkey, rightkey and shootkey however, you have to use SDLK numbers (see link in 1.4.6.)
You might want to tweak controlmaxspeed, controlaccel and controlstartspeed,
which are also explained in 1.2.6.

1.4.3 Digital Joystick
A digital joystick is a stick connected to 4 buttons, therefore it provides
very poor control, ie. its either ON or OFF, therefore, if you use such a joystick
be sure to choose "Digital Joystick" in the joystick options menu. You can tweak
tweak controlmaxspeed, controlaccel and controlstartspeed. (explained in 1.4.6.)

1.4.4 Analog Joystick
A analog joystick provides accurate control, if you use a analog joystick you
need to choose "Analog Joystick" in the joystick options menu, then make sure
the stick is centered, press "Callibrate", push joystick button shortly,then
push the stick left, press button shortly, and finally push the stick right and press
button shortly.
You can tweak controlmaxspeed (explained in 1.4.6.) it acts as sensitivity.
Callibration values are saved to config so you won't need to callibrate again.

1.4.5 WiiMote
You need to compile sdl-ball with support for libwiiuse, read more about this under 2.2.
You need to connect the wiimote every time you start the game.
To connect the wiimote simply press "w" then press 1 and 2 on the WiiMote.
You can run sdl-ball from a terminal to see some debugging message from libwiiuse.
You can tweak controlmaxspeed (explained in 1.4.6) it acts as sensitivity.

1.4.6 Config options explained
clearscreen [0/1] - clears the scene before drawing (looks pretty funny if you dont)
eyecandy [0/1] - turns on nice particle effects and trails, saves a few cpu cycles to turn it off
resx - horizontal resolution
resy - vertical resolution
showbg [0/1] - show the background image
fullscreen [0/1] - use full-screen mode
particlecollide [0/1] - use collision detection for particles, looks nice but might a cpu hog
sound [0/1] - use sound effects (sound effects will allways be loaded) (requires compiled with sound support)
stereo [0/1] - use stereo, some people don't like stereo, might save a few cpu cycles to turn it off
controlmaxspeed - the max speed the paddle can travel with when using non-mouse controllers.
                  10 is VERY fast, this acts as sensitivity for WiiMote and analog joystick.
controlaccel - the acceleration that the paddle gains when using digital joystick or keyboard for controlling.
controlstartspeed - the starting speed (the "lowest" speed) of the paddle when using digital joystick or keyboard.
rightkey - SDLK number of the key that makes the paddle travel right
leftkey - same as above, but left.
nextkey - Select next powerup.
buykey  - Buy selected powerup.
prevkey - Select previous powerup.
shootkey - the key that makes the guns shoot and releases the ball.
Check this site for sdl key numbers: http://kobesearch.cpan.org/htdocs/SDL_Perl/SDL/Constants.pm.html#SDLK_0-
joyisdigital [0/1] - if this is 1, the joystick acts as a "dump"/digital joystick (behaving excatly like the keyboard)
                     if this is 0, the joystick acts as a analog/precise joystick.
joycal* - callibration values from the joystick callibration process.
joyenabled [1/0] - if 1 then joystick is enabled (set to 0 if you experience random movement/controlling error)
sndtheme - Case sensetive name of the sound-theme dir to use.
gfxtheme - Case sensetive name of the gfx-theme dir to use.
lvltheme - Case sensetive name of the level-theme dir to use.

------------------------------------------------------------------------------------------------------------

2.0 Compiling
The basic requirements for compiling sdl-ball is the following libs AND THEIR DEVELOPMENT FILES:
OpenGL
LibGLU
SDL1.2
SDL-image1.2
SDL-ttf1.2
SDL-mixer1.2
Make, a working C++ compiler and stdlibc++6

To compile, extract the SDL-Ball tarball (since you're reading this I figure you allready did) and cd into the dir.
Type make and press enter.

If you get errors: error: SDL.h: No such file or directory
And you're SURE you have dev libs installed, add -I/usr/include/SDL to CFLAGS in Makefile

2.1 Without sound
If you don't have a soundcard/have a very weak computer/don't have SDL-mixer you can compile without sound support.
This removes SDL-mixer dependancy.
To turn off sound support, you need to edit two files:
Makefile: remove -lSDL_mixer
main.cpp: remove #define WITH_SOUND

2.2 With WiiMote support
To enable WiiMote support you need to have a working installation of
libwiiuse ( http://www.wiiuse.net/ ) I won't get into installing that.
You also need a working blue-tooth dongle w/driver and all, if you can't run the
example-program from wiiuse, you need to fix those problems before trying to compile sdl-ball.
You need to have a systemwide install of wiiuse ( sudo make install in the wiiuse dir ).
You need to edito two files to enable WiiMote support:
Makefile: add -lwiiuse to LDFLAGS
main.cpp: remove the // in front of #define WITH_WIIUSE

2.3 For systemwide installation Because every linux dist is diffrent, 
I can't know where it would be nice to put the datafiles (themes) for sdl-ball,
you might want to use a dir like /usr/local/share/sdl-ball
In that case, make a directory there, and put 
the themes dir into it. Then compile with
make -DDATADIR="/usr/share/sdl-ball/themes"

------------------------------------

3.0 Themes
Themes can, but don't have to, consist of several things, in SDL-Ball a theme can contain:
levels - other levelsets are always nice.
graphics - all the graphics in sdl-ball can be changed by the theme.
sounds - you can customise sounds aswell.
Themes can be stored in 2 diffrent places: Either in ~/.config/sdl-ball/themes/
or in sdl-balls own sdl-ball/themes folder.

3.1 How to use a diffrent theme
You can choose a theme from the options menu.
You can mix sound, gfx and level thems in the settings file (see 1.4 and 1.4.6)

3.2 How to make your own theme
http://sdl-ball.sourceforge.net/?page=themes for more information on themes, it's really easy! ;)

------------------------------------

4.0 Contributors and Credits
People who have directly helped the SDL-Ball project, chronologically:
Quinton "quin" Reeves - Windows Port.
Ciddk - Website and Bugreports
Evgeni Golov - Suggestions, comments, insight, Debian package
"Dionos" - The dio-sound-theme.
yugrotavele - The AUR package
kfgz - Patching
Yannickescu - AROS port


------------------------------------

5.0 Frequently Asked Questions
Q: The paddle moves around randomly, or is stuck in one of the sides?
A: try disabling joystick support in the settings.cfg


changelog.txt
0.01
Initial alpha release, highscores are broken, there is a few bugs in the collision detection, and performance is bad.

0.02
Initial highscore support

0.03
Replaced all timer code with two global variables, instead of every object keeping track of its own time which only uses additional resources and makes the game out of sync.
Minor fixes
Fixed the "lives left" text.
Fixed bug in highscores so it only showed up first time
Added settings.cfg and support for loading and saving settings.

0.03b
Fixed bug that made it impossible to change level, added a nonpausing tick variable
Added background to the highscore listing

0.03c
Cleaned up the menu class.
Made highscore listing look better.

0.04
Using lists for effects and balls instead, for more effective management of collision detection.
Optimized a bit here and there.

0.05
Optimizing overall, code cleanup, not drawing the screen every time the simulation runs
implemented global tick variables for frame-age.
implemented aged frames, so the screen is now refreshed in every loop of the engine, major performance boost
game will now write/use settings.cfg to remember settings if available (it will try to create it when a setting is changed in the menu)
particle collision detection is off by default (turn it on in settings.cfg)
fixed missing texture in sound-options
game should now be optimized enough to be playable on most modern machines
removed joystick init from sdl because it is not used

0.05b
fixing non-bug where a ball hits two bricks at the same time, and changing direction because hit detection is done at the first brick, and then at the second brick, it detects the ball as hitting the brick on another surface than what whould see realistic (this is not really a bug, but it looks better.)
removing unused -lglut from makefile, it is not used.

0.05c
Optimizing a bit
Made aim user controlled, using the mousewheel to aim.
Made level editor plugin print out ** Start ** and ** Stop **.

0.06
Added speedometer
Tweaked difficulty settings and speed-increment values
Points are calculated on the average ballspeed, so if the balls are moving faster, more points is given.
Made the game restart when all levels are completed, with a bigger multiplyer on points
Stuff I don't remember
Don't render invisible bricks

0.07
Announced text is in front of menu/dialogs
Announce-class cleanup, fixed bug with messages getting shown wrong if they were right after eachother.
Announce only lifes left and levelchanges. Not powerups.
Removed moving backgrounds.
Added internal fps counter.
Fixed bug in multi-ball powerup that only made it spawn one ball.
Game now restores resolution on exit.
Aim is no longer user-controlled - made game too easy.
Fixed speed bug
Fixing bug that made glass-bricks break on first hit.

0.08
Texture manager can now load jpg images, but only if the filename ends in jpg (case sensitive)
Using jpg background images now.
Clean up bullets on screen when changing level
Changes to levels.txt about random powerups.
Moved powerup properties to a seperate file (data/powerups.txt)
Some powerups are more rare than others
extra life powerup
Remove edge bleeding from ball/paddle-extensions
TTF rendering is VERY heavy.
  Using prerendered numbers for the highscore instead.
  File data/gfx/highscores/numbers.png and template added 
show lives left
Make sure there is at least one explosive powerup in level 4
dont announce lives left since its visible now.
ball tail color is changed instant when "explosive" powerup applies.
rewrote random-powerup code, 4 classes of powerups (most (common) 60%, more 30%, less 10%, least 1%)
initial try to balance powerup spawning.
Make a "random evil powerup" function.
Put a random evil poweups on Doombricks.

0.08b
Found that both P and 3 was assigned to GUN powerup, only P is used now.
Javascript leveleditor added, makes editing levels a lot easier.
Remade level 3 and 11
Added level 20,21
Removed data-changelog.
Thanks to David Wührer for pointing out a (the??) bug in the collision detection.

0.09
bricks can now drop/lift and rotate left or right (moveBoard(brick bricks[], int dir))
Die powerup
Add particle effect to paddle when player dies.
Powerup objects on screen explodes when player looses life.
Added evil powerup that makes board drop down one brick-height when ball hits paddle.
Make 5% chance of getting a powerup default for all levels instead of 10%.
Added "Detonate all explosive bricks" powerup
Added level options for scolling levels, see data/levels.txt for info
Add Explosive brick grow powerup.
Games can now be saved and loaded.
Saving is only available in the beginning of levels, and not in level 1.
Corrected bug that made brick spawn powerup/effects if it is hit twice in one frame.
New (quick/easy) level 1 (moved old level 1 to level 4)
More levels.


0.10
Will choose current resolution if resx or resy is not found in settings.cfg.
Added "All bricks can be destroyed" powerup (PO_EASYBRICK)
Collision detection improved. (ball wont deflect off two sides at the same time, which was annoying!)
Next-level powerup added (player won't get score for bricks that are not hit this level).
Slow/Fast ball poweups wont be added, since there is allready "Small balls"/"Normal balls" and "Big Balls" powerups that do about the same thing.
Aim helper pwerup added (when a ball is on collision course with paddle, a line is drawn to show deflection angle)
Fixed bug that made PO_EXPLOSIVE_GROW happen upon ball collision instead of powerup collision.. xD
Puts user-files (highscore,savegame,settings) in ~/.config/sdl-ball
Added remove rule to Makefile to delete ~/.config/sdl-ball
Only ask for name if player has a high score.
Improved deflection precision.
Added support for padded textures.
Give player aimhelper when theres one brick left.
New powerup icons and other gfx.

0.11
README added.
Removed "flash" at exit that comes from trying to restore resolution even if it is run in windowed mode.

loadlevel.cpp
/* ************************************************************************* * 
    SDL-Ball - DX-Ball/Breakout remake with openGL and SDL for Linux 
    Copyright (C) 2008 Jimmy Christensen ( dusted at dusted dot dk )

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 * ************************************************************************* */

class powerupLoaderClass {
  public:
  //most common
  string chances[4];
  int powerups;
  int powerupsGiven;
  string evilPowerups;

  powerupLoaderClass() {
    int chance=-1;
    powerups=0;
    powerupsGiven=0;
    string line;
    
    //Which powerups are evil?
    evilPowerups = "2H37"; //P
    
    ifstream powerFile(useTheme("/powerups.txt",setting.lvlTheme).data());
    if(!powerFile.is_open())
    {
      cout << " Could not open 'powerups.txt'"<<endl;
      var.quit=1;
      return;
    }
      
    while(!powerFile.eof())
    {
      getline(powerFile, line);
      if(line.compare("Most:")==0)
      {
        chance=0;
      } else if(line.compare("More:")==0)
      {
        chance=1;
      } else if(line.compare("Less:")==0)
      {
        chance=2;
      } else if(line.compare("Least:")==0)
      {
        chance=3;
      } else {
        //No matches, either its a comment or a powerup.
        if(line.compare(0,2,"//")!=0 && line.compare("")!=0)
        {
          chances[chance] += line[0];
          powerups++;
        }
      }

    }
  }

  char selectRandomPowerup()
  {
    int i = rand()%100;
    powerupsGiven++;
    if(i < 1) //1% of powerups are from this class
    {
      i=rand()%chances[3].length();
      return(chances[3][i]);
    } else if(i < 11) //10% of powerups are from this class
    {
      i=rand()%chances[2].length();
      return(chances[2][i]);
    } else if(i < 41) //30% of powerups are from this class
    {
      i=rand()%chances[1].length();
      return(chances[1][i]);
    } else { //Rest of the powerups are from this class
      i=rand()%chances[0].length();
      return(chances[0][i]);
    }
  }

  char randomPowerup(char p)
  {
    int i;
    //First, decide if we are going to put a powerup in the brick. (based on p)
    i=rand()%100;
    if(p=='Q') //100% chance of evil
    {
      return(randomEvilPowerup());
    } else if(p=='K') //100% chance
    {
      return(selectRandomPowerup());
    } else if(p=='N' && i < 1) //1%
    {
     return(selectRandomPowerup());
    } else if(p=='M' && i < 2) //2%
    {
      return(selectRandomPowerup());
    } else if(p=='L' && i < 5) //5%
    {
      return(selectRandomPowerup());
    } else if(p=='J' && i < 10) //10%
    {
      return(selectRandomPowerup());
    }
   
    //Return no powerup or return the specified.
    if(p=='K' || p=='N' || p=='M' || p=='L' || p=='J')
    {
      return('0');
    } else {
      return(p);
    }

  }
  
  char randomEvilPowerup()
  {
    int i=rand()%evilPowerups.length();
    return(evilPowerups[i]);
  }

};

void loadlevel(string file, brick bricks[] ,int level)
{
  ifstream levelfile(file.data());
  if(!levelfile.is_open())
  {
    cout << " Could not open " << file << endl;
    var.quit=1;
    return;
  }
  string line;
  int levelread=0,brick=0,ch=0;
  var.numlevels=0;
  var.scrollInfo.drop = 0;

  
  while(!levelfile.eof())
  {
    getline(levelfile,line);
    //Har vi fundet start?
    if(levelread == 0)
    {
      //Nej, er det nu?
      if(line.substr(0,11) == "** Start **")
      {
        levelread++;
      }
    } else if(levelread == 1)
    {
      //Do the level stop now?
      if(line.substr(0,10) == "** Stop **")
      {
        levelread=0;
        var.numlevels++;
        brick=0;
      } else { //Reading data from level
        
        if(var.numlevels == level)
        {
          if(line[0] == '>')
          {
            if(line.substr(0,6) == "> down")
            {
              var.scrollInfo.dropspeed = atol( line.substr(7,line.length()).data() );
              var.scrollInfo.drop = 1;
            }
          } else {
            while(line[ch] != 0)
            {
              bricks[brick].powerup = line[ch];
              bricks[brick].type = line[ch+1];
              
              if(bricks[brick].type == 'D')
              {
                char rgb[3][5];
                
                sprintf(rgb[0], "0x%c%c", line[ch+2], line[ch+3]);
                sprintf(rgb[1], "0x%c%c", line[ch+4], line[ch+5]);
                sprintf(rgb[2], "0x%c%c", line[ch+6], line[ch+7]);
                
                bricks[brick].tex.prop.glTexColorInfo[0] = 0.003921569*strtol(rgb[0], NULL, 16);
                bricks[brick].tex.prop.glTexColorInfo[1] = 0.003921569*strtol(rgb[1], NULL, 16);
                bricks[brick].tex.prop.glTexColorInfo[2] = 0.003921569*strtol(rgb[2], NULL, 16);
                
                bricks[brick].tex.prop.glParColorInfo[0] = 0.003921569*strtol(rgb[0], NULL, 16);
                bricks[brick].tex.prop.glParColorInfo[1] = 0.003921569*strtol(rgb[1], NULL, 16);
                bricks[brick].tex.prop.glParColorInfo[2] = 0.003921569*strtol(rgb[2], NULL, 16);
                
                bricks[brick].tex.prop.glTexColorInfo[3] = 1.0;
                ch +=6;
              }
              //cout << "Level: " << levelnum  << " brick: " << brick << " Powerup: " << line[ch] << " Type: " << line[ch+1]<<"\n";
  
              brick++;
              ch +=2;
            }
            ch=0;
          } //Not a command
        } //denne level settes op
      }
    }
  }
   cout << "Read " << var.numlevels << " levels from '"<< file <<"'"  << endl;
  levelfile.close();
}

void initlevels(brick bricks[], textureClass texLvl[])
{
  int brick;
  powerupLoaderClass powerupLoader;

  //Temp storage for custom colors
  GLfloat tempCol[4], tempParCol[3];

  //Set dem op
  int row,i;
  i=0;

  for(row=0;row<23;row++)
  {

    for(brick=0;brick < 26; brick++)
    {
    
      bricks[i].posx =  -1.54 + (GLfloat)brick*0.1232;
      bricks[i].posy =  0.95+ ((GLfloat)row*0.07)*-1;
      
      if(bricks[i].type != '0')
      {
        bricks[i].active=1;
        bricks[i].collide=1;
        bricks[i].isdyingnormally=0;
        bricks[i].isexploding=0;
        bricks[i].reflect=1;

        bricks[i].fade=1.0;
        bricks[i].fadespeed=2.0;
        bricks[i].zoom=1.0;
        bricks[i].zoomspeed=4.0;
        bricks[i].width=0.0616;
        bricks[i].height=0.035;
        
        bricks[i].score=11;
        bricks[i].destroytowin=1;
        bricks[i].hitsLeft=1;
             
        bricks[i].powerup = powerupLoader.randomPowerup(bricks[i].powerup);

        bricks[i].bricknum=brick;

        bricks[i].row=row;
        updated_nbrick[row][brick] = i; // This brick is active
        //cout<<"Brick:"<< nbrick[row][brick]<<endl;
      }  else {
        bricks[i].score=0;
        bricks[i].destroytowin=0;
        bricks[i].active=0;
        updated_nbrick[row][brick]=-1;
       // cout<<"Brick:"<< nbrick[row][brick]<<endl;
      }
        

      
      
      if(bricks[i].type == '1')
      {
        bricks[i].tex=texLvl[6];
        bricks[i].opacity=texLvl[6].prop.glTexColorInfo[3];
      } else if(bricks[i].type == '2')
      {
        bricks[i].tex=texLvl[7];
        bricks[i].opacity=texLvl[7].prop.glTexColorInfo[3];
      }else if(bricks[i].type == '3')
      {
        bricks[i].powerup = '0';
        bricks[i].tex= texLvl[2];
        bricks[i].opacity=texLvl[2].prop.glTexColorInfo[3];
        bricks[i].score=30;
        bricks[i].destroytowin=0;
      } else if(bricks[i].type == '4')
      {
        bricks[i].tex=texLvl[4]; //glass texture
        bricks[i].hitsLeft = 2; //takes two hits to kill
        bricks[i].opacity=texLvl[4].prop.glTexColorInfo[3];
      } else if(bricks[i].type == '5')
      {
        bricks[i].tex=texLvl[8];
        bricks[i].opacity=texLvl[8].prop.glTexColorInfo[3];

      
      } else if(bricks[i].type == '6')
      {
        bricks[i].tex=texLvl[9];
        bricks[i].opacity=texLvl[9].prop.glTexColorInfo[3];

      
      } else if(bricks[i].type == '7')
      {
        bricks[i].tex=texLvl[10];
        bricks[i].opacity=texLvl[10].prop.glTexColorInfo[3];

      
      } else if(bricks[i].type == '8')
      {
        bricks[i].tex=texLvl[11];
        bricks[i].opacity=texLvl[11].prop.glTexColorInfo[3];

      } else if(bricks[i].type == '9')
      {
        bricks[i].tex=texLvl[5]; //invisible texture
        bricks[i].tex.frame=1;
        bricks[i].hitsLeft = 3; //takes 3 to kill
        bricks[i].opacity=texLvl[5].prop.glTexColorInfo[3];
      
      } else if(bricks[i].type == 'A')
      {
        bricks[i].tex=texLvl[12];
        bricks[i].opacity=texLvl[12].prop.glTexColorInfo[3];
      } else if(bricks[i].type == 'B')
      {
        bricks[i].tex=texLvl[0];
        bricks[i].opacity=texLvl[0].prop.glTexColorInfo[3];
      } else if(bricks[i].type == 'C')
      {
        bricks[i].tex=texLvl[3];
        bricks[i].opacity=texLvl[3].prop.glTexColorInfo[3];
        bricks[i].powerup = powerupLoader.randomEvilPowerup();
      } else if(bricks[i].type == 'D') //type D get colors applied by loadlevel.
      {
        //Hold the colors while applying texture props
        memcpy(tempCol, bricks[i].tex.prop.glTexColorInfo, sizeof(tempCol));
        memcpy(tempParCol, bricks[i].tex.prop.glParColorInfo, sizeof(tempParCol));

        bricks[i].tex = texLvl[1];
        
        //Put pack the colors
        memcpy(bricks[i].tex.prop.glTexColorInfo,tempCol, sizeof(tempCol));
        memcpy(bricks[i].tex.prop.glParColorInfo,tempParCol,sizeof(tempParCol));

        bricks[i].opacity = tempCol[3];
        
      }

      i++;
    }
  }
//    cout << "Powerups given to this level: " << powerupLoader.powerupsGiven << endl;
}
